File: exercism/cpp/anagram/anagram.cpp
#include "anagram.h"
#include <boost/algorithm/string.hpp>

std::string key(const std::string &given) {
  std::string normalized = boost::algorithm::to_lower_copy(given);
  std::sort(normalized.begin(), normalized.end());
  return normalized;
}

std::vector<std::string>
anagram::anagram::matches(std::vector<std::string> candidates) {
  std::string target = key(subject);
  std::vector<std::string> matching;
  for (const auto &candidate : candidates) {
    if (key(candidate) == target &&
        boost::algorithm::to_lower_copy(candidate) !=
            boost::algorithm::to_lower_copy(subject))
      matching.push_back(candidate);
  }
  return matching;
}

File: exercism/go/anagram/anagram.go
package anagram

import "strings"

// Detect selects anagrams of a given word among candidates
func Detect(subject string, candidates []string) (matches []string) {
	sLower := strings.ToLower(subject)
	sCounts := counts(sLower)
	for _, other := range candidates {
		oLower := strings.ToLower(other)
		if oLower == sLower {
			continue
		}
		if counts(oLower) == sCounts {
			matches = append(matches, other)
		}
	}
	return matches
}

func counts(s string) string {
	freqs := make([]byte, 26)
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c < 'a' || c > 'z' {
			continue
		}
		freqs[c-'a']++
	}
	return string(freqs)
}
